src/main/java/lambdastreams/Book.java

```diff
@@ -0,0 +1,199 @@
+                .mapToObj(book -> new Book(generateRandomName(), generateRandomPrice()))
+                .collect(Collectors.toList());
+    }
+
+    public static List<String> getBooksNames(List<Book> allBooks) {
+        return allBooks.stream()
+                .filter(Objects::nonNull)
+                .map(book -> book.getName().trim())
+                .filter(name -> !name.isEmpty())
+                .collect(Collectors.toList());
+    }
+
+    public static BigDecimal getTotalPriceOfAllBooks(List<Book> allBooks) {
+        return allBooks
+                .stream()
+                .map(Book::getPrice)
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+    }
+
+    public static List<String> getBooksNamesWithAString(List<Book> allBooks, String aString) {
+
+        Predicate<String> isContainAString = bookName -> bookName.toLowerCase().contains(aString.toLowerCase());
+
+        return allBooks.stream()
+                .filter(Objects::nonNull)
+                .map(book -> book.getName().trim())
+                .filter(isContainAString)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * There can be few books with the same highest price, so we return LIST of their names
+     *
+     * @param allBooks
+     * @return
+     */
+    public static List<String> getBookWithHighestPrice(List<Book> allBooks) {
+
+        List<BigDecimal> listOfPrices = allBooks.stream()
+                .filter(Objects::nonNull)
+                .map(Book::getPrice)
+                .collect(Collectors.toList());
+
+        Predicate<Book> isMaxPrice = bookName -> bookName.getPrice().compareTo(Collections.max(listOfPrices)) == 0;
+
+        List<Book> listOfBooksWithHighestPrice = allBooks.stream()
+                .filter(Objects::nonNull)
+                .filter(isMaxPrice)
+                .collect(Collectors.toList());
+
+        return getBooksNames(listOfBooksWithHighestPrice);
+    }
+
+    public static int getNumberOfBooksSpecificLength(List<Book> allBooks, int nameLength) {
+
+        Predicate<Book> isSpecificLength = bookName -> bookName.getName().length() == nameLength;
+
+        return allBooks.stream()
+                .filter(Objects::nonNull)
+                .filter(isSpecificLength)
+                .map(Book::getName)
+                .collect(Collectors.toList())
+                .size();
+    }
+
+    public static List<Book> getBooksWithPriceLessThanLimit(List<Book> allBooks, double limit) {
+
+        Predicate<Book> isLessThanLimit = book -> book.getPrice().compareTo(BigDecimal.valueOf(limit)) < 1;
+
+        return allBooks.stream()
+                .filter(Objects::nonNull)
+                .filter(isLessThanLimit)
+                .collect(Collectors.toList());
+    }
+
+    public static List<Book> getListOfBooksSortedDescByName(List<Book> allBooks) {
+
+        return allBooks.stream()
+                .sorted(Comparator.comparing(Book::getPrice))
```

Comment:
it works correctly, but just a hint - you can also use thenComparing() method for sorted() - .sorted(Comparator.comparing(Book::getPrice).thenComparing(Comparator.comparing(Book::getName).reversed()))

Reply:
done